src/FileStorageBase/Invoker/ContractInvoker.cs

```diff
@@ -0,0 +1,152 @@
+using System;
+using System.Linq;
+using Akka.Actor;
+using Neo.Cryptography.ECC;
+using Neo.IO;
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.SmartContract;
+using Neo.SmartContract.Native;
+using Neo.VM;
+using Neo.Wallets;
+
+namespace Neo.FileStorage.Invoker
+{
+    public abstract class ContractInvoker
+    {
+        public Wallet Wallet { get; init; }
+        public NeoSystem NeoSystem { get; init; }
+        public IActorRef Blockchain { get; init; }
+        private WalletAccount account;
+        protected WalletAccount DefaultAccount
+        {
+            get
+            {
+                if (account is null)
+                {
+                    foreach (WalletAccount wa in Wallet.GetAccounts())
+                    {
+                        if (wa.IsDefault)
+                        {
+                            account = wa;
+                            break;
+                        }
+                        if (account == null) account = wa;
+                    }
+                }
+                return account;
+            }
+        }
+
+        protected void Invoke(UInt160 contractHash, string method, long fee, params object[] args)
+        {
+            InvokeResult result = TestInvoke(contractHash, method, args);
+            using SnapshotCache snapshot = NeoSystem.GetSnapshot();
+            uint height = NativeContract.Ledger.CurrentIndex(snapshot);
+            Random rand = new();
+            Transaction tx = new()
+            {
+                Version = 0,
+                Nonce = (uint)rand.Next(),
+                Script = result.Script,
+                ValidUntilBlock = height + NeoSystem.Settings.MaxValidUntilBlockIncrement,
+                Signers = new Signer[] { new Signer() { Account = DefaultAccount.ScriptHash, Scopes = WitnessScope.Global } },
+                Attributes = Array.Empty<TransactionAttribute>(),
+            };
+            tx.SystemFee = result.GasConsumed + fee;
+            tx.NetworkFee = Wallet.CalculateNetworkFee(snapshot, tx);
+            var balance = NativeContract.GAS.BalanceOf(snapshot, DefaultAccount.ScriptHash);
+            if (tx.SystemFee + tx.NetworkFee > balance)
+            {
+                Utility.Log(nameof(ContractInvoker), LogLevel.Debug, $"insufficient gas, network={NeoSystem.Settings.Network}, method={method}, height={height}, system_fee={tx.SystemFee}, network_fee={tx.NetworkFee}, balance={balance}");
+                throw new InvalidOperationException($"make transaction failed, gas insufficient");
+            }
+            var data = new ContractParametersContext(snapshot, tx, NeoSystem.Settings.Network);
+            if (!Wallet.Sign(data)) throw new InvalidOperationException($"make transaction failed, wallet could not sign");
```

Comment:
Maybe we should log the `ContractParametersContext` in order to allow to use MultiSign wallets